## 29190410 TIL

### .java와 class의 관계
1. name.java 파일안에 여러개의 클래스가 있다면, 이름이 name인 클래스만 public이어야 한다.
2. main을 가지는 class 가 있는경우, 자동으로 그 클래스가 public이 된다 
3. 1.과 2.의 경우가 겹쳐서, name.java안의 클래스 중 name클래스가 아닌 다른 클래스에 main이 있다면 순환적인 컴파일에러가 발생한다
4. 특별한 경우가 아니면 원래 이렇게 .java안에 여러 클래스를 만들지 않는다


### 쓸수 있고 없고

|           | class | field | method | local variable |
|:---------:|:-----:|:-----:|:------:|:--------------:|
|   public  |   O   |   O   |    O   |        X       |
| protected |   X   |   O   |    O   |        X       |
|  default  |   O   |   O   |    O   |        X       |
|  private  |   X   |   O   |    O   |        X       |
|   final   |   O   |   O   |    O   |        O       |
|   static  |   X   |   O   |    O   |        X       |
|  abstract |   O   |   X   |    O   |        X       |

### Exception

- 예외 : 프로그램 실행 중 발생할 수 있는 예기치 않은 사건

#### 예외의 종류 

1. Throwable
	1. error : 치명적 오류. 해결불가한 오류
	2. exception
		1. Runtime exception(unchecked exception) : 실행중에 발생하는 예외
			1. arithmetic
			2. out of index
		2. Unhandled exception(checked exception) : 컴파일 전에 handling 해 두어야 하는 exception
			1. IO
			2. DateFormat

- Runtime exception을 방어하는 것에 신경을 많이 써야 한다.

#### 처리
- `try ~ catch~finaly`
- 예외가 발생할 수 있는 구문을 try에서 실행하고, 예상되는 exception을 cath구문에서 받아서 처리해주는 문법
- finaly구문에 들어간 부분은 항상 꼭 실행된다
- 여러개의 exception을 catch할 때는 상속관계 가장 아래부터 올라가는 순으로 쓴다

```java

try{
			new Integer("3.4");
		//	상속관계를 거슬러 올라가는 순으로 작성
		}catch(NumberFormatException e) {
			
		}catch(IllegalArgumentException e) {
			
		}catch(RuntimeException e) {
			
		}catch(Exception e) {
			
		}finally{

		}

```

#### 회피
- `throws` : method의 parameter 옆에 위치하고, return 과 유사하게 exception을 반환한다. 호출시에 반환받은 exception을 처리해야 함

#### 유발
- `throw` : 특정한 exception을 직접 발생시킴 
- `unreachable` : 예외 발생으로 프로그램이 종료됐으니까 이 코드는 실행되지 못한다


#### IO Exception
- 입출력 관련 Exception은 handling을 해주지 않는다. IO에 관련된 제어권은 JVM이 가지고 있지 않고, OS에 있기 때문.
- try를 쓰지 않고, Throws를 사용해서 넘겨버리면 된다.

#### System.err.println();
- 빨갛게 에러메세지로 출력하라


#### 예외처리에 대한 조치
- `System.exit(0);` : 프로그램 자체를 종료시키는 명령 
	- 보통의 경우에는, main메소드에서 흐름에 맞게 종료시키는 것이 좋다. 사용을 지양해야 하는 코드
- `return;` : catch 구문이 위치한 메서드를 종료
- `finally` : 반드시 실행되는 구문
- `try~catch~finally 와 return;` : finally가 return 앞에서 반드시 실행됨
- `try~catch~finally 와 System.exit(0);` : 강제 종료되므로 finall도 실행이 안됨
- 모든 catch마다 return으로 종료할 필요는 없다. 종료하지 않고 상황에 알맞는 조치로 종료하지 않고 작업을 이어갈 수 있다. 
- exception이 throws 되지도 않고, 메서드가 종료되지도 않았을 때는, call한 부분에서는 메소드 내부에서 정상 작업이 됐는지 exception이 발생했는지 알 수 없으므로, **exception을 handling하는 method에서는 bool등의 return값을 주어 exception의 발생 유무를 외부에 전달할 수 있도록 하자**

---
### 오늘 과제

1. ** 4/11까지 ** 컴퓨터로 말고, 손으로 258을 2진수/8진수/16진수 바꾸자

2. ** 4/11까지 ** 5d(16) 를 10진수, 8진수, 2진수로 바꾸자 (손으로)

1. **4/12 오후4시까지** 진수 변경하는 프로그램 자바로 로직 구현(toBinary, toHex 같은 method 쓰지 않고)

```

입력하려는 값을 선택하세요
1. 10진수 2. 2진수 3. 8진수 4. 16진수

OOO를 입력하세요 : XXX
	=>myException 만들어서 원하는 값인지 아닌지 핸들링

XXX는	2진수 : OOO
		8진수 : OOO
		16진수 : OOO
입니다.


```

3. sort 정렬, search 검색
	1. **4/12 오전까지** 버블 소트 원리, 방법
	2. **4/12 오전까지** 순차 소트 원리, 방법
	3. **4/15까지** 이진 트리 검색 원리
	4. java로 sort method 없이(for 사용) 구현
